データベース設計

aiueo818 opened the pull request:
# What
ER図の作成、READMEへの記載

# Why
データベース設計のため

aiueo818 committed:
Revert "Revert "データベース設計""

This reverts commit a6b4a29aeac20f3ce4bb039332e120ddb9723441.

aiueo818 commented:
ER図のGyazoURLは以下になります。よろしくお願いいたします。
https://gyazo.com/7f600211e65e96169dfa2ece753e9c87

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| Column     | Type       | Options     |
+| ---------- | ---------- | ----------- |
+| nickname   | string     | null: false |
+| mail       | string     | null: false |
```
```ruby
| mail | string | null: false |
```
こちらのカラム名はdeviseで用意されている`email`に変更しましょう！
また、重複を避けるため、オプションに`unique: true`をつけましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| ---------- | ---------- | ----------- |
+| nickname   | string     | null: false |
+| mail       | string     | null: false |
+| password   | string     | null: false |
```
```ruby
| password | string | null: false |
```

こちらのカラム名はdeviseで用意されている`encrypted_password`を使用しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| nickname   | string     | null: false |
+| mail       | string     | null: false |
+| password   | string     | null: false |
+| birth      | string     | null: false |
```
```ruby
| birth | string | null: false |
```
生年月日のデータ型は`date`を指定しましょう！

minachanchan replied:
以下のカラムに関しても追加しましょう！
```
・姓
・名
・姓のフリガナ
・名のフリガナ
```

minachanchan replied:
ユーザーと紐づくテーブルは商品情報テーブルと購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■ユーザーと商品情報
ユーザー：親
商品情報：子
1対多で指定

■ユーザーと購入履歴
ユーザー：親
購入履歴：子
1対多で指定
```

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+
+| Column       | Type       | Options                        |
+| ------------ | ---------- | ------------------------------ |
+| image        | references | null: false, foreign_key: true |
```
```ruby
| image | references | null: false, foreign_key: true |
```

画像データはActiveStorageを使用して保存するため、
商品情報テーブルに画像データを保存するカラムは不要でございます。
こちらのカラムは削除しましょう。

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| Column       | Type       | Options                        |
+| ------------ | ---------- | ------------------------------ |
+| image        | references | null: false, foreign_key: true |
+| text         | string     | null: false                    |
```
```ruby
| text | string | null: false |
```
textはrailsの予約語ですので、予期せぬエラーの要因となる恐れがございます。
別の単語のカラム名に修正しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| status       | string     | null: false                    |
+| postage      | string     | null: false                    |
+| address      | references | null: false, foreign_key: true |
+| days         | string     | null: false                    |
```
```ruby
| category     | string     | null: false                    |
| status       | string     | null: false                    |
| postage      | string     | null: false                    |
| address      | references | null: false, foreign_key: true |
| days         | string     | null: false                    |
```

こちらはプルダウンで選択する項目のカラムかとお見受けします。
active_hashでの実装に際し、以下についてご確認をお願いいたします！

①
active_hashを使用して保存するデータは数値を扱うので、データ型はintegerを指定しましょう！


②
active_hashを使用して保存するデータのカラムは、「単数系_id」という形にする必要がございます。
以下のように各カラム名を修正しましょう！

```
例：category_id
```

③
「day」はrailsの予約語ですので、こちらも別の単語を用いたカラム名に変更しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| postage      | string     | null: false                    |
+| address      | references | null: false, foreign_key: true |
+| days         | string     | null: false                    |
+| price        | string     | null: false                    |
```
```ruby
| price | string | null: false |
```

商品価格は数値で扱うので、データ型はintegerを指定しましょう！

minachanchan replied:
商品説明カラム並びにユーザー情報テーブルの外部キーカラムを追記しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| price        | string     | null: false                    |
+
+### Association
+- has_one :buy
```
商品情報と紐づくテーブルはユーザーテーブルと購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■商品情報とユーザー
ユーザー：親
商品情報：子
1対多で指定

■商品情報と購入履歴(50行目で指定済み)
商品情報：親
購入履歴：子
1対1で指定
```

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+
+| Column     | Type       | Options                        |
+| ---------- | ---------- | ------------------------------ |
+| list       | references | null: false, foreign_key: true |
```
```ruby
| list | references | null: false, foreign_key: true |
```
購入履歴テーブルに保存するデータは、「どのユーザー」が「どの商品」を購入したかです。
こちらのカラムは削除し、
「どのユーザー」かはユーザーテーブルの外部キーとして、
「どの商品」かは商品情報テーブルの外部キーとして、カラムを指定しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+
+
+### Association
+- belongs_to :items
```
購入履歴と紐づくテーブルはユーザーテーブル、商品情報テーブル、発送先住所テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■購入履歴とユーザー
ユーザー：親
購入履歴：子
1対多で指定

■購入履歴と商品情報
商品情報：親
購入履歴：子
1対1

■購入履歴と発送先住所
購入履歴：親
発送先住所：子
1対1
```

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| Column         | Type       | Options     |
+| -------------- | ---------- | ----------- |
+| post_code      | string     | null: false |
+| prefecture     | string     | null: false |
```
```ruby
| prefecture | string | null: false |
```

都道府県を保存するカラムでは、商品情報テーブルで使用したactive_hashのデータを兼用できます。
商品情報で指定したものと同じものをこちらでも指定しましょう！

minachanchan commented on README.md:
```diff
@@ -22,3 +22,50 @@ Things you may want to cover:
+| city           | string     | null: false |
+| street_address | string     | null: false |
+| building       | string     |             |
+| phone_number   | string     | null: false |
```
購入履歴テーブルの外部キーカラムを追加しましょう！

minachanchan replied:
発送先住所と紐づくテーブルは購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■発送先住所と購入履歴
購入履歴：親
発送先住所：子
1対1
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

実装完了まで引き続き、一緒に頑張りましょう！
何卒よろしくお願いいたします。

aiueo818 committed:
db設計の修正

# What
レビュー依頼でいただいた修正箇所を修正。ER図の再作成。

# Why
正しくデータベース設計をするため

aiueo818 commented:
ER図のGyazoURLは以下になります。よろしくお願いいたします。
https://gyazo.com/532400ff141e829d284020b0689c6432

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+
+### Association
+- has_many :items
+- has_many :buy
```
```
- has_many :buy
```
has_manyを使用する場合、モデル名は複数形にしましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+- has_many :items
+- has_many :buy
+
+## itemsテーブル
```
実装要件と比較して商品テーブルのカラム数が少ないです。
実装要件及び見本アプリの出品ページを参考にカラムを追加しましょう。

maimai0401 replied:
ユーザーの外部キーカラムを追加しましょう。

■理由
実装要件において、商品テーブルには出品者が誰であるかを保存することが求められているためになります。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+| postage_id        | integer    | null: false, foreign_key: true |
+| address_id        | integer    | null: false, foreign_key: true |
+| number_of_days_id | integer    | null: false, foreign_key: true |
+| price             | integer    | null: false, foreign_key: true |
```
```
| postage_id        | integer    | null: false, foreign_key: true |
| address_id        | integer    | null: false, foreign_key: true |
| number_of_days_id | integer    | null: false, foreign_key: true |
| price             | integer    | null: false, foreign_key: true |
```
カラムのオプションから外部キー制約を外しましょう。
他テーブルと関係を持つ外部キーカラムではないためです。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+| number_of_days_id | integer    | null: false, foreign_key: true |
+| price             | integer    | null: false, foreign_key: true |
+
+### Association
```
アソシエーションの記述に誤りがあります。
適切な記述に修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+- has_many :users
+- belongs_to :buy
+
+## buyテーブル
```
buyテーブルのテーブル名を複数形にしましょう。

■理由
命名規則においてテーブル名は複数形であることが求められているためです。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+| Column     | Type       | Options                        |
+| ---------- | ---------- | ------------------------------ |
+| users      | references | null: false, foreign_key: true |
+| items      | references | null: false, foreign_key: true |
```
```
| users      | references | null: false, foreign_key: true |
| items      | references | null: false, foreign_key: true |
```
カラム名は単数形にしましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+| items      | references | null: false, foreign_key: true |
+
+
+### Association
```
必要なアソシエーションを追記しましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+
+
+### Association
+- belongs_to :address
```
アソシエーションの記述に誤りがあります。
適切な記述に修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+### Association
+- belongs_to :address
+
+## addressテーブル
```
テーブル名は複数形にしましょう。

maimai0401 replied:
必要なカラムが不足しているようです。
追記しましょう。

maimai0401 replied:
アソシエーションを追記しましょう。

maimai0401 commented on README.md:
```diff
@@ -22,3 +22,60 @@ Things you may want to cover:
+| city           | string     | null: false, foreign_key: true |
+| street_address | string     | null: false, foreign_key: true |
+| building       | string     | foreign_key: true              |
+| phone_number   | string     | null: false, foreign_key: true |
```
```
| city           | string     | null: false, foreign_key: true |
| street_address | string     | null: false, foreign_key: true |
| building       | string     | foreign_key: true              |
| phone_number   | string     | null: false, foreign_key: true |
```
カラムのオプションから外部キー制約を外しましょう。
他テーブルと関係を持つ外部キーカラムではないためです。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください。
どうぞよろしくおねがいします。

aiueo818 committed:
DB設計の修正２

# What
レビューいただいた内容を元にER図とREADMEを修正

# Why
DB設計のため

aiueo818 commented:
ER図のURLは以下になります。よろしくお願いいたします。
https://gyazo.com/fb2d734e6312df50c8c5b0f4ab05dc43

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| first_name                 | string     | null: false               |
+| last_name_reading          | string     | null: false               |
+| first_name_reading         | string     | null: false               |
+| buy_id                     | string     | null: false               |
```
◯カラムを削除しましょう！

不要な内容であるため

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+- has_many :items
+- has_many :buys
+
+## itemsテーブル
```
◯カラムを追加しましょう！

現状、一部必要なラムの記載が見られない状況です。
以下をご参照の上、必要なカラムを追加しましょう。


| 商品名        | 文字列型       | NOT NULL 制約                |
| 価格          | 整数型        | NOT NULL 制約                |
| 商品の説明     | テキスト型     | NOT NULL 制約                |
| カテゴリー     | 整数型        | NOT NULL 制約                |
| 商品の状態     | 整数型        | NOT NULL 制約                |
| 配送料の負担   | 整数型        | NOT NULL 制約                |
| 発送元の地域   | 整数型        | NOT NULL 制約                |
| 発送までの日数  | 整数型        | NOT NULL 制約                |
| ユーザー      | 外部キーカラム  | NOT NULL 制約, 外部キーオプション |

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+
+### Association
+- has_one :buys
+- belongs_to :users
```
◯アソシエーションを修正しましょう！

belongs_toの対象は単数形です。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| price             | integer    | null: false |
+
+### Association
+- has_one :buys
```
◯アソシエーションを修正しましょう！

has_oneの対象は単数形です。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| user       | references | null: false, foreign_key: true |
+| item       | references | null: false, foreign_key: true |
+
+### Association
```
◯アソシエーションを追加しましょう！

配送先住所情報テーブルへの内容が必要です。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| Column         | Type       | Options           |
+| -------------- | ---------- | ----------------- |
+| post_code      | string     | null: false       |
+| address_id     | string     | null: false       |
```
◯カラムの型を修正しましょう！

アクティブハッシュカラムはid管理のため、integer型を用います。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| address_id     | string     | null: false       |
+| city           | string     | null: false       |
+| street_address | string     | null: false       |
+| building       | string     | foreign_key: true |
```
◯オプションを修正しましょう！

建物カラムにオプションは不要です。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| street_address | string     | null: false       |
+| building       | string     | foreign_key: true |
+| phone_number   | string     | null: false       |
+| card           | string     | null: false       |
```
◯カラムを冊書しましょう！

cardカラムは不要な内容であるため

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| building       | string     | foreign_key: true |
+| phone_number   | string     | null: false       |
+| card           | string     | null: false       |
+
```
◯カラムを追加しましょう！

購入履歴への外部キーカラムが必要です。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,64 @@ Things you may want to cover:
+| card           | string     | null: false       |
+
+### Association
+- belongs_to :users
```
◯アソシエーションを修正しましょう！

親テーブルは購入履歴テーブルで。

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。


引き続き実装頑張っていきましょう！

```
※注意※
DB設計にて修正箇所のございます際は、最新のER図のキャプチャ画像も改めて添付頂きますようお願い致します。
```

aiueo818 committed:
DB設計の修正３

# What
レビューいただいた箇所の修正（ER図の更新とREADMEの記載）

# Why
DB設計を正しく行うため

ER図URL
https://gyazo.com/209fd0a96f3939f655c4a038d0079b3a

aiueo818 commented:
# What
レビューいただいた箇所の修正（ER図の更新とREADMEの記載）

# Why
DB設計を正しく行うため

ER図URL
https://gyazo.com/209fd0a96f3939f655c4a038d0079b3a

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,67 @@ Things you may want to cover:
+### Association
+- has_many :items
+- has_many :buys
+- has_many :addresses
```
◯アソシエーションを削除しましょう！

has_many :addresses

不要な内容であるため

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,67 @@ Things you may want to cover:
+| Column            | Type       | Options                        |
+| ----------------- | ---------- | ------------------------------ |
+| product_name      | string     | null: false                    |
+| explanation       | integer    | null: false                    |
```
◯カラムの方を修正しましょう！

説明カラムはテキスト型です

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,67 @@ Things you may want to cover:
+| status_id         | integer    | null: false                    |
+| postage_id        | integer    | null: false                    |
+| address_id        | integer    | null: false                    |
+| number_of_days_id | integer    | null: false                    |
```
◯カラム名を修正しましょう！

アクティブハッシュカラムのカラム名は単数形_idとなるため

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,67 @@ Things you may want to cover:
+- has_many :buys
+- has_many :addresses
+
+## itemsテーブル
```
◯カラムを追加しましょう！

https://github.com/aiueo818/furima-40136/pull/12/files#r1433619187

上記をご参照の上、不足しているカラムを追加しましょう。

ShotaOtake commented on README.md:
```diff
@@ -22,3 +22,67 @@ Things you may want to cover:
+| buy_id         | string     | null: false       |
+
+### Association
+- belongs_to :user
```
◯アソシエーションを削除しましょう！

belongs_to :user

不要な内容であるため

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

引き続き実装頑張っていきましょう！

```
※注意※
DB設計にて修正箇所のございます際は、最新のER図のキャプチャ画像も改めて添付頂きますようお願い致します。
```

aiueo818 committed:
DB設計の修正４

# What
レビューいただいた修正箇所の修正

# Why
DB設計を正しく行うため

ER図のURLは以下になります。
https://gyazo.com/f9fbc9ddcc952fcde8fcdb6dcae0f47d

aiueo818 commented:
ER図のURLは以下になります。
https://gyazo.com/f9fbc9ddcc952fcde8fcdb6dcae0f47d

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| Column            | Type       | Options                        |
+| ----------------- | ---------- | ------------------------------ |
+| product_name      | string     | null: false                    |
+| explanation       | string     | null: false                    |
```
４９行目
explanationカラムが商品説明カラムの場合、長い文章が入ることを想定してtext型にしましょう。

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| address_id        | integer    | null: false                    |
+| number_of_day_id  | integer    | null: false                    |
+| price             | integer    | null: false, foreign_key: true |
+| user_id           | integer    | null: false, foreign_key: true |
```
５６行目
外部キーはreferences型を使用して記述しましょう。
（66行目と同様の記述に修正しましょう）

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| street_address | string     | null: false       |
+| building       | string     |                   |
+| phone_number   | string     | null: false       |
+| buy_id         | string     | null: false       |
```
８４行目
こちらも外部キーのため、references型を使用した記述に修正しましょう。
references型の場合は、カラム名の末尾に_idは不要になります。また外部キー制約として、foreign_key: trueの記述が必要になります。

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| postage_id        | integer    | null: false                    |
+| address_id        | integer    | null: false                    |
+| number_of_day_id  | integer    | null: false                    |
+| price             | integer    | null: false, foreign_key: true |
```
５５行目
priceカラムのオプションから外部キー制約を外しましょう。
priceカラムは他テーブルと関係を持つ外部キーカラムではないためです。

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| category_id       | integer    | null: false                    |
+| status_id         | integer    | null: false                    |
+| postage_id        | integer    | null: false                    |
+| address_id        | integer    | null: false                    |
```
５３行目
カラム名を変更しておきましょう。
こちらはActiveHashを利用するカラム名と思われますが、配送先情報テーブル（addressesテーブル）と名前が重複するため、別のカラム名にしておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -22,3 +22,66 @@ Things you may want to cover:
+| Column         | Type       | Options           |
+| -------------- | ---------- | ----------------- |
+| post_code      | string     | null: false       |
+| address_id     | integer    | null: false       |
```
７９行目
53行目を修正いただきましたら、こちらも同じ名前に修正しておきましょう。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

データベース設計は、大変難しいフェーズですが重要な設計になりますので、もう後一息がんばりましょう！

修正後は、再度レビュー依頼フォームのご提出をお願いいたします。
どうぞよろしくお願いいたします！

aiueo818 committed:
DB設計の修正５

# What
レビューいただいた箇所の修正

# Why
DB設計と正しく行うため

ER図のURL
https://gyazo.com/229b3dadcf7b32fca1977541f260611f

aiueo818 commented:
ER図のURLは以下になります。よろしくお願いいたします。
https://gyazo.com/229b3dadcf7b32fca1977541f260611f

TakashiKozuma commented:
追加のレビュー依頼ありがとうございます！
修正点も含め、確認が完了しました。
LGTMです！！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。

aiueo818 commented:
ありがとうございます！！！
